Data/HashMap/Base.hs

```diff
@@ -588,89 +627,135 @@ two = go
         | hy == h = if ky == k
-                    then return $! Leaf h (L k (f x y))
-                    else return $! collision h l (L k x)
-        | otherwise = two s h k x hy ky y
+                    then return $! (0, Leaf h (L k (f x y)))
+                    else return $! (1, collision h l (L k x))
+        | otherwise = do
+              twoHm <- two s h k x hy ky y
+              return $! (1, twoHm)
     go h k x s t@(BitmapIndexed b ary)
         | b .&. m == 0 = do
             ary' <- A.insertM ary i $! Leaf h (L k x)
-            return $! bitmapIndexedOrFull (b .|. m) ary'
+            return $! (1, bitmapIndexedOrFull (b .|. m) ary')
         | otherwise = do
             st <- A.indexM ary i
-            st' <- go h k x (s+bitsPerSubkey) st
+            (sz, st') <- go h k x (s+bitsPerSubkey) st
             A.unsafeUpdateM ary i st'
-            return t
+            return (sz, t)
       where m = mask h s
             i = sparseIndex b m
     go h k x s t@(Full ary) = do
         st <- A.indexM ary i
-        st' <- go h k x (s+bitsPerSubkey) st
+        (sz, st') <- go h k x (s+bitsPerSubkey) st
         A.unsafeUpdateM ary i st'
-        return t
+        return (sz, t)
       where i = index h s
     go h k x s t@(Collision hy v)
-        | h == hy   = return $! Collision h (updateOrSnocWith f k x v)
+        | h == hy   =
+            let !start = A.length v
+                !newV = updateOrSnocWith f k x v
+                !end = A.length newV
+            in return $! (end - start, Collision h newV)
         | otherwise = go h k x s $ BitmapIndexed (mask hy s) (A.singleton t)
-{-# INLINABLE unsafeInsertWith #-}
+{-# INLINABLE unsafeInsertWithInternal #-}
 
 -- | /O(log n)/ Remove the mapping for the specified key from this map
 -- if present.
 delete :: (Eq k, Hashable k) => k -> HashMap k v -> HashMap k v
-delete k0 m0 = go h0 k0 0 m0
+delete k0 (HashMap sz m0) =
+    let (diff, m0') = deleteInternal k0 m0
+    in HashMap (sz + diff) m0'
+{-# INLINABLE delete #-}
+
+-- | /O(log n)/ Remove the mapping for the specified key from this map
+-- if present. Returns a tuple with the hashmap's change in size and the
+-- hashmap after the deletion.
+deleteInternal :: (Eq k, Hashable k) => k -> HashMapInner k v -> (Int, HashMapInner k v)
+deleteInternal k0 m0 = go h0 k0 0 m0
   where
     h0 = hash k0
-    go !_ !_ !_ Empty = Empty
+    go !_ !_ !_ Empty = (0, Empty)
     go h k _ t@(Leaf hy (L ky _))
-        | hy == h && ky == k = Empty
-        | otherwise          = t
+        | hy == h && ky == k = (-1, Empty)
+        | otherwise          = (0, t)
     go h k s t@(BitmapIndexed b ary)
-        | b .&. m == 0 = t
+        | b .&. m == 0 = (0, t)
         | otherwise =
             let !st = A.index ary i
-                !st' = go h k (s+bitsPerSubkey) st
-            in if st' `ptrEq` st
+                (!sz, !st') = go h k (s+bitsPerSubkey) st
+            in (sz,) (if st' `ptrEq` st
```

Comment:
Why all the laziness? Did you mean `!(!sz, !st')`? And did you mean `(sz,) $! ...`?

Reply:
The first one was unintentional, I first wrote `!(sz, st)` but then realized that wouldn't force each element of the tuple. TIL I should do both.